Sources/MessagingPush/PushClickHandler.swift

```diff
@@ -0,0 +1,220 @@
+                    deepLinkUtil.handleDeepLink(deepLinkUrl)
+                }
+            }
+        default: break
+        }
+    }
+
+    func trackMetric(
+        notificationContent: UNNotificationContent,
+        event: Metric
+    ) {
+        guard let deliveryID: String = notificationContent.userInfo["CIO-Delivery-ID"] as? String,
+              let deviceToken: String = notificationContent.userInfo["CIO-Delivery-Token"] as? String
+        else {
+            return
+        }
+
+        trackMetric(deliveryID: deliveryID, event: event, deviceToken: deviceToken)
+    }
+
+    // There are files that are created just for displaying a rich push. After a push is interacted with, those files
+    // are no longer needed.
+    // This function's job is to cleanup after a push is no longer being displayed.
+    func cleanupAfterPushInteractedWith(pushContent: CustomerIOParsedPushPayload) {
+        pushContent.cioAttachments.forEach { attachment in
+            let localFilePath = attachment.url
+
+            try? FileManager.default.removeItem(at: localFilePath)
+        }
+    }
+}
+
+// UNUserNotificationCenterDelegate functions
+extension PushClickHandlerImpl {
+    // Notification was interacted with.
+    @objc public func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
+        _ = MessagingPush.shared.userNotificationCenter(center, didReceive: response, withCompletionHandler: completionHandler)
+    }
+}
+
+// Swizzle functions
+extension PushClickHandlerImpl {
+    // Swizzled method that gets called when a new UNUserNotificationCenter.delegate gets set.
+    @objc dynamic func cio_swizzled_setDelegate(delegate: UNUserNotificationCenterDelegate?) {
+        guard let delegate = delegate else {
+            cio_swizzled_setDelegate(delegate: delegate) // continue swizzle
+            return
+        }
+
+        setupClickHandling(onDelegate: delegate)
+
+        cio_swizzled_setDelegate(delegate: delegate) // continue swizzle
+    }
+
+    // Swizzled method that gets called when a push notification gets clicked on
+    @objc dynamic func cio_swizzle_didReceive(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
+        _ = MessagingPush.shared.userNotificationCenter(center, didReceive: response, withCompletionHandler: completionHandler)
+
+        // TODO: duplicate metrics could be reported because we might be swizzling multiple delegates and so this function gets called X number of times.
+        // https://github.com/customerio/issues/issues/11150 should we do this?
+
+        // continue swizzle
+        cio_swizzle_didReceive(center, didReceive: response, withCompletionHandler: completionHandler)
+    }
+}
+
+// TODO: cleanup duplicate swizzle functions all over codebase.
```

Comment:
To avoid scope creep, cleaning up of duplicate swizzled functions will be done in another PR. 

Work has begun on branch `levi/duplicate-swizzling`